### ISSUE TITLE
Apollo observers and subscriptions not properly handled if component is recreated

### ISSUE THREAD

**Issue by dimitri-koenig** [Akryum/vue-apollo#229]
Hi there

This may be related to #223, but in this case it's not an SSR error.

Expected: When using vue-router and switching between components, vue-apollo should follow the destruction of a component on route-change and destroy all observer handlers and subscriptions.

Current state:
On recreation apollo may use already existing data from cache and thus call the result handler even before the vue created() hook is finished.
Additionally the loading state does not work properly anymore on component recreation, very strange.

Any idea?

**Comment by Kocal** [comment]
I agree, it's really annoying when developing a SPA with vue-router. :confused: 

The error `TypeError: Cannot read property 'currentResult' of null` ([full trace](https://gist.github.com/Kocal/81b1c508f2e72a72e4a6b217a39de5c1#file-logs-L77-L87)) keeps being thrown each time I go from page **B** to page **A**, then coming back to page **B**.

If it can help mainteners, below you will find an example of my component which is contained inside a page component. 
When the page is created, this component is created, when the page is destroyed, the component is destroyed too.

```js
// MyComponent.js

import { mapGetters } from 'vuex';
import MyGraphQLQuery from '../foo.graphql';

export default {
  data() {
    return {
      loading: 0,
      fooBar: null,
    }
  },
  computed: {
    ...mapGetters(['site']),
  },
  apollo: {
    fooBar: {
      manual: true,
      query: MyGraphQLQuery,
      variables() {
        return {
          siteId: this.site.id,  
        } 
      },
      result({ data, loading }) {
        if (!loading) {
          this.fooBar = data.site.fooBar;
        }
      },
    }
  }
}
```

Thanks

**Comment by dohomi** [comment]
@dimitri-koenig @Kocal  did you try wrapping your data inside the result hoos as follows:
```
result({data}){
   this.$nextTick(() => {
    this.fooBar = data.site.fooBar
  })
}
```

**Comment by dimitri-koenig** [comment]
Yes, tried that. result() is called before even the created() hook of a component. Thus the error. nextTick doesn't solve that issue for me.

**Comment by Akryum** [comment]
Note: if you use the default fetchPolicy, and you already have your data in the case, it is expected that data is immediately rendered and no call to the API is made.

**Comment by Akryum** [comment]
Could someone provide a reproduction or a failing test in this repo (in a PR)?

**Comment by Kocal** [comment]
I can't do that now, maybe this weekend if I got time :/
